File: app/src/main/java/com/absinthe/kage/protocol/ProtocolHandler.java
package com.absinthe.kage.protocol;

import android.util.Log;

import com.absinthe.kage.connect.tcp.KageSocket;
import com.absinthe.kage.device.Device;
import com.absinthe.kage.device.cmd.InquiryDeviceInfoCommand;
import com.absinthe.kage.device.model.DeviceConfig;

public class ProtocolHandler implements IProtocolHandler {
    private static final String TAG = ProtocolHandler.class.getSimpleName();
    private Device mDevice;
    private DeviceConfig mConfig;
    private IProtocolHandleCallback mCallback;
    private boolean hasHandShake = false;

    public ProtocolHandler(Device device, DeviceConfig config, IProtocolHandleCallback callback) {
        this.mDevice = device;
        this.mConfig = config;
        this.mCallback = callback;
    }

    @Override
    public void handleSocketConnectedEvent() {
        InquiryDeviceInfoCommand inquiryDeviceInfoCommand = new InquiryDeviceInfoCommand();
        inquiryDeviceInfoCommand.phoneName = mConfig.name;
        mDevice.sendCommand(inquiryDeviceInfoCommand);
    }

    @Override
    public void handleSocketMassage(String msg) {
        if (msg != null) {
            String[] split = msg.split(IpMessageProtocol.DELIMITER);
            try {
                int cmd = Integer.parseInt(split[0]);

                switch (cmd) {
                    case IpMessageConst.GET_CLIENTTYPE:
                        hasHandShake = true;
                        KageProtocolThreadHandler.getInstance().post(() -> {
                            if (null != mCallback) {
                                mCallback.onProtocolConnected();
                            }
                        });
                        break;
                    default:
                }
            } catch (Exception e) {
                Log.e(TAG, "protocol invalid:" + e.getMessage());
            }
        }
    }

    @Override
    public void handleSocketDisConnectEvent() {
        if (!hasHandShake) {// 握手期间断开了连接判定为连接失败
            KageProtocolThreadHandler.getInstance().post(() -> {
                if (null != mCallback) {
                    mCallback.onProtocolConnectedFailed(KageSocket.ISocketCallback.CONNECT_ERROR_CODE_HAND_SHAKE_UNCOMPLETE, null);
                }
            });
            return;
        }
        hasHandShake = false;
        KageProtocolThreadHandler.getInstance().post(() -> {
            if (null != mCallback) {
                mCallback.onProtocolDisConnect();
            }
        });
    }

    @Override
    public void handleSocketConnectFail(final int errorCode, final Exception e) {
        KageProtocolThreadHandler.getInstance().post(() -> {
            if (null != mCallback) {
                mCallback.onProtocolConnectedFailed(errorCode, e);
            }
        });
    }

    @Override
    public void handleSocketSendOrReceiveError() {
        KageProtocolThreadHandler.getInstance().post(() -> {
            if (null != mCallback) {
                mCallback.onProtocolSendOrReceiveError();
            }
        });
    }
}

File: app/src/main/kotlin/com/absinthe/kage/connect/protocol/ProtocolHandler.kt
package com.absinthe.kage.connect.protocol

import com.absinthe.kage.connect.protocol.IProtocolHandler.IProtocolHandleCallback
import com.absinthe.kage.connect.tcp.KageSocket
import com.absinthe.kage.device.Device
import com.absinthe.kage.device.cmd.InquiryDeviceInfoCommand
import com.absinthe.kage.device.model.DeviceConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import timber.log.Timber

class ProtocolHandler(
        private val mDevice: Device,
        private val mConfig: DeviceConfig,
        private val mCallback: IProtocolHandleCallback) : IProtocolHandler {

    private var hasHandShake = false

    override fun handleSocketConnectedEvent() {
        val inquiryDeviceInfoCommand = InquiryDeviceInfoCommand().apply {
            phoneName = mConfig.name
        }
        mDevice.sendCommand(inquiryDeviceInfoCommand)
    }

    override fun handleSocketMassage(msg: String) {
        val split = msg.split(IpMessageProtocol.DELIMITER).toTypedArray()
        try {
            when (split[0].toInt()) {
                IpMessageConst.GET_DEVICE_INFO -> {
                    hasHandShake = true
                    GlobalScope.launch(Dispatchers.Main) {
                        mCallback.onProtocolConnected()
                    }
                }
            }
        } catch (e: Exception) {
            Timber.e("protocol invalid: $e")
        }
    }

    override fun handleSocketDisConnectEvent() {
        if (!hasHandShake) {
            // 握手期间断开了连接判定为连接失败
            GlobalScope.launch(Dispatchers.Main) {
                mCallback.onProtocolConnectedFailed(KageSocket.ISocketCallback.CONNECT_ERROR_CODE_HAND_SHAKE_NOT_COMPLETE, null)
            }
            return
        }
        hasHandShake = false
        GlobalScope.launch(Dispatchers.Main) {
            mCallback.onProtocolDisConnect()
        }
    }

    override fun handleSocketConnectFail(errorCode: Int, e: Exception) {
        GlobalScope.launch(Dispatchers.Main) {
            mCallback.onProtocolConnectedFailed(errorCode, e)
        }
    }

    override fun handleSocketSendOrReceiveError() {
        GlobalScope.launch(Dispatchers.Main) {
            mCallback.onProtocolSendOrReceiveError()
        }
    }
}